### PR TITLE
CNV-21826: Configuring VM for DPDK

### PR DESCRIPTION
**Pull request by sjhala-ccs**
Version(s): 4.13+

Issue: [CNV-21826](https://issues.redhat.com//browse/CNV-21826)

Link to docs preview: https://58584--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.html#virt-configuring-cluster-dpdk_virt-attaching-vm-to-sriov-network

QE review:
- [x] QE has approved this change.

### CODE REVIEW

File: modules/virt-configuring-vm-dpdk.adoc
Comment (apinnick) [inline modules/virt-configuring-vm-dpdk.adoc:61]: Just a small nit, since I will be fixing these in a separate PR. The OCP guidelines say to use `...` instead of `#...` for ellipses in YAML files.

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:69]:
```suggestion
<7> The name of the SR-IOV `NetworkAttachmentDefinition` object.
```

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:77]: The values use here and in the following commands need to match some of the attributes earlier used in the VM spec, used in other commands of this section (and maybe even in the PerformanceProfile). We should call those out. @orelmisan would you please help defining the mapping?

**Review by phoracek — CHANGES_REQUESTED** [review]

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (orelmisan) [reply modules/virt-configuring-cluster-dpdk.adoc:20]: @phoracek , I am not aware of a difference between a `PerformanceProfile` for Pods and a `PerformanceProfile` for VMs.

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (apinnick) [inline modules/virt-configuring-cluster-dpdk.adoc:27]: I think line 27 has an extra indent. Shouldn't it align with line 26?

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (apinnick) [inline modules/virt-configuring-cluster-dpdk.adoc:19]: Small nit. I think this should be "an", not "a", because AFAIK "SR-IOV" is an initialism (pronounced "ess-ar-eye-oh-vee" etc), not an acronym.

File: virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.adoc
Comment (apinnick) [inline virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.adoc:32]: I don't think you need this link because the module is included in this assembly.

File: modules/virt-configuring-vm-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-dpdk.adoc:77]: @orelmisan Do you have any suggested changes for this section?

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (phoracek) [reply modules/virt-configuring-cluster-dpdk.adoc:21]: I like that it describes the individual steps clearly. The only issue with this is that it mixes steps that are required for the cluster configuration and those needed for namespace configuration. Could you split the list to make it clear what needs to be done once for the whole cluster and what is required for every new project.

Comment (orelmisan) [reply modules/virt-configuring-cluster-dpdk.adoc:21]: I'll change the place of this step:
> Configure the runtimeClass in the kubevirt CR.

1. Map your compute nodes topology - understand to which NUMA node your SR-IOV NIC is connected to.
2. Label the worker nodes to mark them as DPDK-enabled.
3. Create a MachineConfigPool object to represent the DPDK-enabled worker nodes.
4. Create a PerformanceProfile object with a number of isolated CPUs for your DPDK application from the same NUMA as the NIC (and reserve the rest).
5. Configure the runtimeClass in the kubevirt CR.
6. Create a namespace for your DPDK applications.
7. Create an SriovNetworkNodePolicy object.
8. Create an SriovNetwork object, and configure it to create the NetworkAttachmentDefinition object in the namespace you've just created.
9. Run the VM latency checkup to check that the network is properly configured.
10. Run the DPDK checkup.

Comment (phoracek) [reply modules/virt-configuring-cluster-dpdk.adoc:21]: I did not mean reordering the list, but having two different lists. One would be guide to configure the cluster for cluster admin on day-1, the other would be a guide to configure a namespace for cluster admin (or powerful namespace admin) on day-2

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-cluster-dpdk.adoc:24]:
```suggestion
. Map your compute nodes topology - understand to which NUMA node your SR-IOV NIC is connected to.
. Label a subset of your worker nodes as DPDK-enabled, e.g. `oc label nodes <node_name> node-role.kubernetes.io/worker-dpdk=""`.
. Create a new `MachineConfigPool` applied to the labeled nodes.
. Create a new `PerformanceProfile` applied to the labeled nodes and the previously configured machine config pool. This profile should select `isolated` CPUs for your DPDK application and `reserved` CPUs for house keeping. Refer to the `PerformanceProfile` documentation for more detail.  
. Create an `SriovNetworkNodePolicy` object with the `spec.deviceType` field set to `vfio-pci`.
. Set the container runtime configuration for the `virt-launcher` pod by adding the following annotation to the `HyperConverged` custom resource (CR):
```

XXX: This may have support implications. We are testing an alternative option. If that option will look ok, we will just need to drop this last step and add a note to the "Create a new `PerformanceProfile`"

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:4]: Could we add a new section between cluster and VM configuration?

```suggestion
:_content-type: PROCEDURE
[id="virt-configuring-vm-dpdk_{context}"]
= Configuring a project for DPDK workloads

Prior to running a DPDK-enabled VM, its namespace/project must be configured according.

.Prerequisites
* Your cluster is configured to run DPDK workloads.

.Procedure
. Create a namespace for your DPDK applications.
. Create an `SriovNetwork` object referencing the earlier defined `SriovNetworkNodePolicy`.
. (Optional) Run the VM latency checkup to check that the network is properly configured.
. (Optional) Run the DPDK checkup to verify that the namespace is DPDK-ready.
```

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:13]:
```suggestion
* You have created and configured a project in which the VM will run.
```

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (phoracek) [reply modules/virt-configuring-cluster-dpdk.adoc:21]: We added a suggested changes below. Closing this thread.

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (phoracek) [reply modules/virt-configuring-cluster-dpdk.adoc:23]: Understood, we can close this thread.

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:39]: There is a BZ we discovered after publishing the upstream docs. We can resolve https://bugzilla.redhat.com/show_bug.cgi?id=2184914 it here (or we can handle it in a follow-up PR).

```suggestion
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
```

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:73]:
```suggestion
.. Using the GRUB bootloader command line interface:
... Configure huge pages. In the example below we use 8 1G huge pages. 
... Select isolated VM's CPUs. In the example below we leave the first two (0 and 1) CPUs for house keeping and isolate the rest for the DPDK application.
```

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [reply modules/virt-configuring-vm-dpdk.adoc:77]: Commented in the thread above. We can close this one.

**Review by phoracek — CHANGES_REQUESTED** [review]

File: modules/virt-configuring-vm-dpdk.adoc
Comment (phoracek) [inline modules/virt-configuring-vm-dpdk.adoc:33]: To make sure we land on the correct node

```suggestion
    spec:
      nodeSelector:
        node-role.kubernetes.io/worker-dpdk: ""
```

**Review by phoracek — CHANGES_REQUESTED** [review]

**Comment by sjhala-ccs** [comment]
@phoracek and @orelmisan I have incorporated your suggestions and added some more details. Please review. Thanks!

Preview build: https://58584--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.html#virt-configuring-cluster-dpdk_virt-attaching-vm-to-sriov-network

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (orelmisan) [inline modules/virt-configuring-cluster-dpdk.adoc:81]: After applying these two manifests, there should be a gradual reboot of the relevant nodes.
Please consider adding a note about it.

File: modules/virt-configuring-vm-project-dpdk.adoc
Comment (orelmisan) [inline modules/virt-configuring-vm-project-dpdk.adoc:44]: Please consider adding a `<x>` here to explain that this is the namespace you want to deploy the NetworkAttachmentDefinition object to.

File: modules/virt-configuring-vm-project-dpdk.adoc
Comment (orelmisan) [inline modules/virt-configuring-vm-project-dpdk.adoc:20]: @phoracek Is there a difference between creating a namespace or creating a project?

**Review by orelmisan — COMMENTED** [review]
Thank you for the changes @sjhala-ccs.
Over all it looks very good.
Please see the inline comments.

File: modules/virt-configuring-vm-project-dpdk.adoc
Comment (phoracek) [reply modules/virt-configuring-vm-project-dpdk.adoc:20]: Frankly, I don't know if OpenShift handles the two differently. I would stick to whatever we use elsewhere in the docs

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (phoracek) [reply modules/virt-configuring-cluster-dpdk.adoc:109]: Sounds good (although I still hope to replace it with the performance profile global)

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-dpdk.adoc:115]: I think it renders fine in the preview, the spacing is somehow not showing up correctly in the file view.

Preview:  https://58584--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.html#virt-configuring-cluster-dpdk_virt-attaching-vm-to-sriov-network

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-dpdk.adoc:58]: Okay, but here is the documentation for the PPC in case you are interested: https://docs.openshift.com/container-platform/4.12/scalability_and_performance/cnf-create-performance-profiles.html

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-dpdk.adoc:81]: Does the user initiate the node reboot or does it happen automatically when the MCP and PP manifests are applied?

File: modules/virt-configuring-vm-project-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-project-dpdk.adoc:20]: @orelmisan This is the documentation on projects that I have linked to in the Additional resources section: https://docs.openshift.com/container-platform/4.12/applications/projects/working-with-projects.html

Is it relevant for this procedure?

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-dpdk.adoc:81]: @orelmisan Does this sound okay: 

> Perform a graceful restart of the compute nodes after you apply the `MachineConfigPool` and `PerformanceProfile`manifests.

I can also link to this procedure if it is relevant: https://docs.openshift.com/container-platform/4.9/nodes/nodes/nodes-nodes-rebooting.html#nodes-nodes-rebooting-gracefully_nodes-nodes-rebooting

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (orelmisan) [reply modules/virt-configuring-cluster-dpdk.adoc]: Thank you

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (orelmisan) [reply modules/virt-configuring-cluster-dpdk.adoc:83]: > Does the user initiate the node reboot or does it happen automatically when the MCP and PP manifests are applied?

The nodes are automatically rebooted.
I don't know if the user has the ability to manually control this process.

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (orelmisan) [reply modules/virt-configuring-cluster-dpdk.adoc:117]: There is a redundant backslash that should be removed.
IIRC it breaks the command.

File: modules/virt-configuring-vm-project-dpdk.adoc
Comment (orelmisan) [reply modules/virt-configuring-vm-project-dpdk.adoc:20]: > @orelmisan This is the documentation on projects that I have linked to in the Additional resources section: https://docs.openshift.com/container-platform/4.12/applications/projects/working-with-projects.html
> 
> Is it relevant for this procedure?

I think so, thanks.

**Review by orelmisan — APPROVED** [review]
Thank you for the changes @sjhala-ccs

File: modules/virt-configuring-vm-dpdk.adoc
Comment (yossisegev) [inline modules/virt-configuring-vm-dpdk.adoc:68]: "ReQuest" - I understand that it is written this way in order to clarify the IRQ abbreviation, but are you sure it is valid to write it this way (i.e. CamelCase inside a single word)?
If this is indeed a common practice in technical writing, then I apologize in advance for the noise.

File: modules/virt-configuring-vm-dpdk.adoc
Comment (apinnick) [inline modules/virt-configuring-vm-dpdk.adoc:64]: FYI: I updated YAML ellipses in all virt modules, changing `...` to `# ...`.

File: modules/virt-configuring-vm-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-dpdk.adoc:68]: Thanks @yossisegev. Our style guidelines do say " When you spell out an abbreviation, do not capitalize letters in the middle of a word to show the origin of an initialism."

I will update the text.

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-cluster-dpdk.adoc:9]:
```suggestion
You can use the following procedure to configure an {product-title} cluster to run Data Plane Development Kit (DPDK) workloads.
```

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-cluster-dpdk.adoc:21]: I know in IBM style we try to avoid the future tense and passive voice as much as we can, which this sentence has a bit of. It still reads fine to me, and I can't think of a suggestion, but I just wanted to make you aware 😄.

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-cluster-dpdk.adoc:22]:
```suggestion
. Label a subset of the compute nodes with a custom role. For example, `worker-dpdk`:
```

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-cluster-dpdk.adoc:55]:
```suggestion
. Create a `PerformanceProfile` manifest that applies to the labeled nodes and the machine config pool that you generated. The performance profile specifies the CPUs that are isolated for DPDK applications and the CPUs that are reserved for house keeping.
```

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-cluster-dpdk.adoc:121]: Up to you! 
```suggestion
Adding the annotation to the `HyperConverged` CR changes a global setting that affects all VMs that are created after the annotation applies. 
```

File: modules/virt-configuring-vm-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-vm-dpdk.adoc:16]: This is also up to you. I was just thinking when you edit something, it already exists, and just I thought the positive `your` might sounds better.
```suggestion
. Edit your `VirtualMachine` manifest to include information about the SR-IOV network interface, CPU topology, CRI-O annotations, and huge pages:
```

File: modules/virt-configuring-vm-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-vm-dpdk.adoc:83]: Does this also work for sub versions of RHEL 8? That might be helpful to specify

File: modules/virt-configuring-vm-project-dpdk.adoc
Comment (kelbrown20) [inline modules/virt-configuring-vm-project-dpdk.adoc:9]:
```suggestion
You can configure the project to run DPDK workloads on SR-IOV hardware. 
```

**Review by kelbrown20 — COMMENTED** [review]
Hello Shikha! Fantastic job on this! I just added some small nits and suggestions, let me know if you have any questions. Great work!

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-dpdk.adoc:21]: Reworded to use present tense

File: modules/virt-configuring-cluster-dpdk.adoc
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-dpdk.adoc:121]: Keeping as is

**Comment by sjhala-ccs** [comment]
@yossisegev Can you please approve the changes if they look good? I need to get an "LGTM" from you before I can merge the PR. Thank you.

**Comment by yossisegev** [comment]
LGTM (CNV network QE)

**Pull request merged**
